### PR TITLE
Passing context directly instead of `IBusInterface`

### PR DESCRIPTION
**Pull request by danielmarbach**
@SzymonPobiega here is the PR. 

Talked to @timbussmann and we couldn't come up with a scenario where we would benefit from the `IBusInterface` passed into `FeatureStartupTask` and `IWantToRunWhenTheBusStartsAndStops`. We considered the hurdle to create a context everytime from the end-user perspective more cumbersome than beneficial.

### CODE REVIEW

File: src/NServiceBus.PerformanceTests/Program.cs [hunk 1]
```diff
@@ -134,15 +134,15 @@ static void DumpSetting(string[] args)
                                   args[5]);
         }
         
-        static void SeedSagaMessages(IBusInterface bus, int numberOfMessages, string inputQueue, int concurrency)
+        static void SeedSagaMessages(IBusContext context, int numberOfMessages, string inputQueue, int concurrency)
         {
-            var sendContext = bus.CreateSendContext();
+            var busContext = context;
```
Comment (timbussmann) [inline src/NServiceBus.PerformanceTests/Program.cs]: remove that?

File: src/NServiceBus.PerformanceTests/Program.cs [hunk 2]
```diff
@@ -150,10 +150,10 @@ static void SeedSagaMessages(IBusInterface bus, int numberOfMessages, string inp
             }
         }
 
-        static TimeSpan SeedInputQueue(IBusInterface bus, int numberOfMessages, string inputQueue, int numberOfThreads, bool createTransaction, bool twoPhaseCommit, bool encryption)
+        static TimeSpan SeedInputQueue(IBusContext context, int numberOfMessages, string inputQueue, int numberOfThreads, bool createTransaction, bool twoPhaseCommit, bool encryption)
         {
             var sw = new Stopwatch();
-            var sendContext = bus.CreateSendContext();
+            var busContext = context;
```
Comment (timbussmann) [inline src/NServiceBus.PerformanceTests/Program.cs]: remove that too?

File: src/NServiceBus.PerformanceTests/PubSub/PubSubTestCase.cs [hunk 3]
```diff
@@ -66,9 +66,9 @@ public override void Run()
                 break;
         }
 
-        configuration.RunWhenEndpointStartsAndStops(new StartActionRunner(b =>
+        configuration.RunWhenEndpointStartsAndStops(new StartActionRunner(context =>
         {
-            var sendContext = b.CreateSendContext();
+            var busContext = context;
```
Comment (timbussmann) [inline src/NServiceBus.PerformanceTests/PubSub/PubSubTestCase.cs]: another one

**Comment by timbussmann** [comment]
I like the changes :+1:

**Comment by danielmarbach** [comment]
@SzymonPobiega @timbussmann pushed a few changes. What do you think? See especially [this change](https://github.com/Particular/NServiceBus/commit/b9c56213ddd37490ca1259c6263e0887a4df8fe2)

@andreasohlund Thoughts?

File: src/NServiceBus.Core/Persistence/InMemory/Outbox/InMemoryOutboxPersistence.cs [hunk 4]
```diff
@@ -33,19 +34,22 @@ class OutboxCleaner : FeatureStartupTask
...
             }
 
-            protected override void OnStop()
+            protected override Task OnStop(IBusContext context)
             {
                 using (var waitHandle = new ManualResetEvent(false))
                 {
                     cleanupTimer.Dispose(waitHandle);
 
+                    // TODO: Use async synchronisation primitve
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Persistence/InMemory/Outbox/InMemoryOutboxPersistence.cs]: Why not do this now?

**Comment by andreasohlund** [comment]
I like it :+1:

**Comment by timbussmann** [comment]
looks good to me :+1:

**Comment by danielmarbach** [comment]
@SzymonPobiega @timbussmann can you merge then into the other branch?

**Comment by SzymonPobiega** [comment]
Passing `IBusContext` instead of `IBusInterface` means that if we align the send context with transactions (which IMO makes sense) the startup tasks would not be able to control transactions. Then I assume each task would get its own send context so that they can succeed/fail independently.

**Comment by danielmarbach** [comment]
@SzymonPobiega We could change that behind the scenes. Currently the implementation reuses the context. Are you still Ok with that or should we switch back to pass in `IBusInterface`?

**Comment by timbussmann** [comment]
@SzymonPobiega how would a task control transactions? it can only create new transactions even when passing `IBusInterface`, right? So the only thing it could do, would be having multiple transactions in one feature.Does that make sense? Can't that be done using new TransactionScopes?

**Comment by SzymonPobiega** [comment]
@danielmarbach @timbussmann the autosubscribe task send messages and it needs to send them in multiple transactions because partial failures are a valid case there.

**Comment by danielmarbach** [comment]
@timbussmann I talked to @SzymonPobiega and convinced that we optimize in this release for usable and not for a potential future batch API. So @SzymonPobiega is good with the current approach

**Comment by timbussmann** [comment]
@SzymonPobiega there would still be the option on using DI to inject the endpoint interface and use CreateBusContext as "low level backdoor"

**Pull request merged**